test/puppetlabs/http/client/sync_plaintext_test.clj

```diff
@@ -510,3 +510,187 @@
               response (common/get client "http://localhost:8080/hello" {:as :text})]
           (is (= 302 (:status response)))
           (common/close client))))))
+
+(defmacro timeout-exception-thrown?
+  [& body]
+  `(try
+     (testlogging/with-test-logging ~@body)
+     (throw (IllegalStateException.
+              "Expected HttpClientException but none thrown!"))
+     (catch HttpClientException e#
+       (if-let [cause# (.getCause e#)]
+         (if (instance? SocketTimeoutException cause#)
+           true
+           (throw (IllegalStateException.
+                    (str
+                      "Expected SocketTimeoutException cause but found: "
+                      cause#))))
+         (throw (IllegalStateException.
+                  (str
+                    "Expected SocketTimeoutException but no cause found.  "
+                    "Message: " (.getMessage e#))))))))
+
+(deftest short-connect-timeout-nonpersistent-java-test-sync
+  (testing (str "connection times out properly for non-persistent java sync "
+                "request with short timeout")
+    (let [request-options     (-> "http://127.0.0.255:65535"
```

Comment:
I found it hard to come up with a good way to exercise the connection timeout.  If you try to connect to a port on the local machine, I've found that the timeout is essentially irrelevant - because the local OS is able to determine right away that nothing is listening on the port and so rejects the connection with a Java `ConnectException` right away.  If a connection is made to a remote machine, it's much more likely to consider the full timeout.

I just tried to pick a "remote" ip address and port for which a connection was extremely unlikely to succeed.  I'm open to a better "remote" ip address and port to use for this case.